freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/AbstractUpgradeCcmAction.java

```diff
@@ -47,7 +50,17 @@ protected UpgradeCcmContext createFlowContext(FlowParameters flowParameters, Sta
 
     @Override
     protected Object getFailurePayload(P payload, Optional<UpgradeCcmContext> flowContext, Exception e) {
-        return new UpgradeCcmFailureEvent(UPGRADE_CCM_FAILED_EVENT.event(), payload.getResourceId(), e);
+        Tunnel oldTunnel = null;
+        Class<? extends AbstractUpgradeCcmEventHandler> failureOrigin = null;
+        if (payload instanceof UpgradeCcmFailureEvent) {
+            UpgradeCcmFailureEvent ufe = (UpgradeCcmFailureEvent) payload;
+            oldTunnel = ufe.getOldTunnel();
+            failureOrigin = ufe.getFailureOrigin();
+        } else if (payload instanceof UpgradeCcmEvent) {
```

Comment:
I would create a separate constructor without the failure origin parameter for this scenario as it is set explicitly to `null`, but this is a stylish question so I am also fine with this solution.